#pragma once

#include "iplugin.h"

namespace Ui {
class GitCommandsForm;
}

class GitPlugin : public IPlugin {
    Q_OBJECT
    struct Config {
        CONFIG_DEFINE(GitBinary, QString)
        CONFIG_DEFINE(GitHomepage, QString)
        CONFIG_DEFINE(GitLastCommand, QString)
        CONFIG_DEFINE(GitLastDir, QString)
        CONFIG_DEFINE(GitLastActiveItem, QString)
        qmdiPluginConfig *config;
    };
    Config &getConfig() {
        static Config configObject{&this->config};
        return configObject;
    }

    enum class GitLog { File, Project };

  public:
    GitPlugin();
    ~GitPlugin();

    // IPlugin interface
    virtual void on_client_merged(qmdiHost *host) override;
    virtual void on_client_unmerged(qmdiHost *host) override;
    virtual void loadConfig(QSettings &settings) override;

  public slots:
    void logFileHandler();
    void logProjectHandler();
    void diffFileHandler();
    void revertFileHandler();
    void refreshBranchesHandler();
    void diffBranchHandler();
    void newBranchHandler();
    void deleteBranchHandler();
    void logHandler(GitPlugin::GitLog log, const QString &filename);
    void on_gitCommitClicked(const QModelIndex &mi);
    void on_gitCommitDoubleClicked(const QModelIndex &mi);

  public slots:
    QString runGit(const QStringList &args, bool saveConfig);
    QString detectRepoRoot(const QString &path);
    QString getDiff(const QString &path);
    QString getRawCommit(const QString &sha1);
    void restoreGitLog();

  private:
    QAction *diffFile = nullptr;
    QAction *logFile = nullptr;
    QAction *logProject = nullptr;
    QAction *revert = nullptr;
    QAction *commit = nullptr;
    QAction *stash = nullptr;
    QAction *branches = nullptr;
    QString gitBinary = "git";
    QString repoRoot;
    QDockWidget *gitDock = nullptr;
    Ui::GitCommandsForm *form = nullptr;
};
